File: python/239_SlidingWindowMaximum.py
from unittest import TestCase
# https://leetcode.com/problems/sliding-window-maximum/
import collections


class SlidingWindowMaximum(TestCase):
    def maxSlidingWindow(self, nums, k):
        """
        :type nums: List[int]
        :type k: int
        :rtype: List[int]
        """
        res, maxes = [], collections.deque()
        for i, v in enumerate(nums):
            while maxes and nums[maxes[-1]] < v:
                maxes.pop()
            maxes.append(i)
            if maxes[0] <= i-k:
                maxes.popleft()
            if i >= k-1:
                res.append(nums[maxes[0]])
        return res

    def test1(self):
        self.assertEqual([3,3,5,5,6,7], self.maxSlidingWindow([1,3,-1,-3,5,3,6,7], 3))

    def test2(self):
        self.assertEqual([], self.maxSlidingWindow([], 0))

File: kotlin/src/main/kotlin/239.SlidingWindowMaximum.kt
import org.junit.Assert.assertEquals
import org.junit.Test
import java.util.*

//https://leetcode.com/problems/sliding-window-maximum/


class SlidingWindowMaximum{
  class MaxQue{
    val values = LinkedList<Int>()
    private val maxes = LinkedList<Int>()

    fun max() = maxes.last
    fun add(v: Int){
      while (maxes.isNotEmpty() && maxes.peek() < v)
        maxes.pop()
      maxes.push(v)
      values.push(v)
    }
    fun remove() = values.removeLast()
        .also { if(it == maxes.last) maxes.removeLast() }
  }

  fun maxSlidingWindow(nums: IntArray, k: Int): IntArray {
    if(k==0) return intArrayOf()
    val que = MaxQue()
    val n = nums.size

    for(i in 0 until minOf(n,k))
      que.add(nums[i])

    val result = mutableListOf(que.max())
    for(i in k until n){
      que.remove()
      que.add(nums[i])
      result.add(que.max())
    }
    return result.toIntArray()
  }

  @Test
  fun test1(){
    assertEquals(intArrayOf(3,3,5,5,6,7).toList(), maxSlidingWindow(intArrayOf(1,3,-1,-3,5,3,6,7), 3).toList())
  }

}